extend/component/implementation/python.md

```diff
@@ -111,82 +157,254 @@ with open('in/tables/source.csv', mode='rt', encoding='utf-8') as in_file, open(
+
+# print list of files matching specific tag
+logging.info(input_files_by_tag['my_tag']) 
+
+# group by name
+input_files_by_name = ci.get_input_file_definitions_grouped_by_name(only_latest_files=True)
+
+# print list of files matching specific name
+logging.info(input_files_by_name['image.jpg'])
+
+```
+
+#### Processing state files
+
+[State files](/extend/common-interface/config-file/#state-file) can be easily loaded and written 
+using the `get_state_file()` and `write_state_file()` methods:
+ 
+```python
+from keboola.component import CommonInterface
+from datetime import datetime
+import logging
+
+# init the interface
+ci = CommonInterface()
+
+last_state = ci.get_state_file()
+
+# print last_updated if exists
+logging.info(f'Previous job stored following last_updated value: {last_state.get("last_updated","")})')
+
+# store new state file
+ci.write_state_file({"last_updated": datetime.now().isoformat()})
+```
+
+### Logging
+
+The library automatically initializes STDOUT or GELF logger based on the presence of the `KBC_LOGGER_PORT/HOST` environment variables 
+upon the `CommonInterface` initialization. To use the GELF logger just enable the logger for your application in the 
+[Developer Portal](https://components.keboola.com/). 
+More details about logging options are available in a [dedicated article](/extend/common-interface/logging/#examples).
+
+With either setting, you can log your messages using the logging library:
+
+```python
+from keboola.component import CommonInterface
+from datetime import datetime
+import logging
+
+# init the interface
+ci = CommonInterface()
+
+logging.info("Info message")
+```
+
+You may also choose to override the settings by enabling the GELF or STDOUT explicitly and specifying the host/port parameters:
+
+```python
+from keboola.component import CommonInterface
+import os
+import logging
+
+# init the interface
+ci = CommonInterface()
+os.environ['KBC_LOGGER_ADDR'] = 'localhost'
+os.environ['KBC_LOGGER_PORT'] = 12201
+ci.set_gelf_logger(log_level=logging.INFO, transport_layer='UDP')
+
+logging.info("Info message")
+```
 
-## Logging
-In Python components, the output is buffered, but the buffering may be [switched off](https://stackoverflow.com/questions/107705/disable-output-buffering). The easiest solution is to run your script with the `-u` option: you would use `CMD python -u ./main.py` in your `Dockerfile`.
-See a [dedicated article](/extend/common-interface/logging/#examples) if you want to
-implement a GELF logger.
+If you use STDOUT logging note that in Python components, the output is buffered. The buffering may 
+be [switched off](https://stackoverflow.com/questions/107705/disable-output-buffering). The easiest solution is to run your script
+with the `-u` option: you would use `CMD python -u ./main.py` in your `Dockerfile`.
 
 ## Error Handling
 The following [piece of code](https://github.com/keboola/component-generator/blob/master/templates/python-tests/src/main.py) is a good entry point:
```

Comment:
What with this? This one https://bitbucket.org/kds_consulting_team/kbc-python-template/src/master/src/component.py violates the user/app error separation and this one https://github.com/keboola/component-generator/blob/master/templates/python-tests/src/main.py is not usable, because you're throwing ValueError for datadir/config errors and treats them like user errors.

Reply:
This one https://github.com/keboola/component-generator/blob/master/templates/python-tests/src/main.py  was there before and is not maintained by KDS. We were talking with COMP (@webrouse) about adding support for the new template and reworking the generator a bit.

https://bitbucket.org/kds_consulting_team/kbc-python-template/src/master/src/component.py  we updated this to throw app error on anything that is not handled explicitly. Perhaps I should add some generic UserError exception and catch it in the __main as well.

Reply:
FYI: We wanted to add a new Python template to the component generator, it's a `New Component Framework` initiative. It has a low priority on the product side and we are NOT currently working on it.

Reply:
this is just about providing the correct guidance.

@davidesner - it's a slightly tricky problem, having
```
        except ValueError as e:
            logging.exception(e)
            exit(1)
```
is fine as long as you have that always the same. Once the user exit occurs on multiple places, someone goes and writes exit(5) there or someone goes and prints the stack trace there, the system is gone. 
We have sort of pivoted to the solution that the php framework provides UserException and that is handled gracefully and everything else causes fire. So that means you just convert the exceptions in app code and handle them in one place

Reply:
@odinuv fair point. I updated the template so the UserException is catched in one place https://bitbucket.org/kds_consulting_team/kbc-python-template/src/de77d414e6ec9a29bceaf548e687038823835d90/src/component.py#lines-80